azure-pipelines.yml

```diff
@@ -6,106 +6,58 @@ pool:
-    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
-
-  # Push test images
-  - script: |
-      docker tag local/dfe-teachers-payment-service:test-dependencies $(dockerRegistry)/$(dockerImageName):cache-test-dependencies
-
-      docker push $(dockerRegistry)/$(dockerImageName):cache-test-dependencies
-    displayName: Push test dependencies Docker image for caching
-    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
-  - script: |
-      docker tag local/dfe-teachers-payment-service:test $(dockerRegistry)/$(dockerImageName):cache-test
-
-      docker push $(dockerRegistry)/$(dockerImageName):cache-test
-    displayName: Push test Docker image for caching
-    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
-
-  # Push web images
-  - script: |
-      docker tag local/dfe-teachers-payment-service:web-dependencies $(dockerRegistry)/$(dockerImageName):cache-web-dependencies
-
-      docker push $(dockerRegistry)/$(dockerImageName):cache-web-dependencies
-    displayName: Push web dependencies Docker image for caching
-    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
-  - script: |
-      docker tag local/dfe-teachers-payment-service:web $(dockerRegistry)/$(dockerImageName):$(Build.BuildNumber)
-      docker tag local/dfe-teachers-payment-service:web $(dockerRegistry)/$(dockerImageName):latest
-
-      docker push $(dockerRegistry)/$(dockerImageName):$(Build.BuildNumber)
-      docker push $(dockerRegistry)/$(dockerImageName):latest
-    displayName: Push web Docker image
+stages:
+  - stage: ci_build
+    displayName: Run tests and publish artifacts
+    jobs:
+      - job: run_tests
+        displayName: Run tests
+        steps:
+          # Build and run tests
+          - script: |
+              docker pull $(dockerRegistry)/$(dockerImageName):cache-test-dependencies || true
+              docker pull $(dockerRegistry)/$(dockerImageName):cache-test || true
+
+              docker-compose --file=docker-compose.test.yml build
+            displayName: Build test Docker image
+          - script: docker-compose --file=docker-compose.test.yml run --rm test
+            displayName: Run tests on Docker
+
+          # Clean up
+          - script: |
+              git reset --hard
+              git clean -xdf
+            displayName: Clean repository
+      - job: build_and_push_web_dependencies
+        dependsOn: run_tests
+        displayName: Build and push web dependencies
+        steps:
+          - script: docker login -u $(dockerId) -p $pass
+            env:
+              pass: $(dockerPassword)
+            displayName: Login to DockerHub
+          - script: |
+              ./bin/build-docker-images "ref/heads/master" \
+                $(Build.SourceVersion) \
+                $(dockerRegistry)/$(dockerImageName) \
+                $(Build.BuildNumber)
+            displayName: Prepare and push Docker images
+      - job: publish_repository
```

Comment:
Do we only want to publish the repository when merging to master?

Reply:
Yeah, this just publishes the artifact, which we need for deploy.

Reply:
It doesn't have a `condition` though, so won't it publish the artifact on non-master builds?

Reply:
Oh gotcha. You're right 😄